pageserver/src/page_service.rs

```diff
@@ -357,8 +357,13 @@ impl PageServerHandler {
 
         /* Send a tarball of the latest snapshot on the timeline */
 
-        let req_lsn = lsn.unwrap_or_else(|| timeline.get_last_valid_lsn());
-
+        let req_lsn = match lsn {
```

Comment:
Why is this needed? In send_tarball(), the first call to get_page_at_lsn() or list_nonrels() or similar will implicitly call wait_lsn anyway, right? (If they don't do that, I think we should add wait_lsn() calls to those functions instead).

If we want to do an explicit wait_lsn() call, I think that would belong in `send_tarball`, not here.

Reply:
Not all repository methods are waiting for lsn now, for example list_rls/list_nonrels.
In principle it is possible to add wait_lsn here. But it seems to be much more efficient to call wait_lsn at the beginning of basebackup, rather than repeat this check in access to each object.
I do no see something bad in pblishing wait_lsn method.


> There seem to be a few different changes here:
> 
>     * Handle CopyFailed message
> 
>     * Add the explicit wait_lsn() call
> 
>     * accept an extra basebackup parameter
> 
> 
> I'd suggest splitting this up into multiple PRs.

I extracted handling of copy error in separate PR:
https://github.com/zenithdb/zenith/pull/447

I do not think that extracting of  wait_lsn should be done separate PR because it is needed only for basebackup

Reply:
> Not all repository methods are waiting for lsn now, for example list_rls/list_nonrels.
In principle it is possible to add wait_lsn here. But it seems to be much more efficient to call wait_lsn at the beginning of basebackup, rather than repeat this check in access to each object.
I do no see something bad in pblishing wait_lsn method.

Well, it's a weirdly inconsistent API now. Some functions wait implicitly (get_rel_size, get_page_at_lsn), while others expect that the caller has done the wait already (list_rels, list_nonrels). At the very least that requires comments. But it would be more consistent if they all waited. Or none of them waited.

The base backup code isn't performance critical, a few wait_lsn calls won't make any difference. On performance grounds though, you could argue for removing the wait_lsn calls from all the Timeline interface functions. In principle, the code in page_service.rs could call wait_lsn once, when it receives the first GetPageLsn request, and store that LSN in a local variable. In subsequent requests, it could compare the new request LSN with the previous one, and only call wait_lsn if it has changed. But currently, it doesn't make any difference in practice, there is so much other overhead in every request that the wait_lsn calls don't matter.